packages/@aws-cdk/aws-s3/lib/auto-delete-objects-handler/index.ts

```diff
@@ -48,9 +77,58 @@ async function emptyBucket(bucketName: string) {
   }
 }
 
-async function onDelete(bucketName?: string) {
+async function onDelete(stackId: string, logicalId?: string, bucketName?: string) {
   if (!bucketName) {
     throw new Error('No BucketName was provided.');
   }
-  await emptyBucket(bucketName);
+  if (!logicalId) {
+    throw new Error('No Logical ID was provided.');
+  }
+  if (await isBucketAboutToBeDeleted(stackId, logicalId)) {
+    await emptyBucket(bucketName);
+  }
 }
+
+/**
+ * Go and inspect CloudFormation to see if the target bucket is about to be deleted
+ */
+async function isBucketAboutToBeDeleted(stackId: string, logicalId: string) {
+  const stackResponse = await cfn.describeStacks({ StackName: stackId }).promise();
+  if (!stackResponse.Stacks?.[0]) {
+    throw new Error(`Could not find stack with ID: ${stackId}`);
+  }
+  const stackStatus = stackResponse.Stacks[0].StackStatus;
+  process.stdout.write(`Stack status: ${stackStatus}\n`);
+
+  // Case 1: the stack failed creation.
+  // Case 2: the stack is being deleted.
+  // In both cases, by definition the bucket will go bye-bye.
+  if (stackStatus === 'ROLLBACK_IN_PROGRESS' || stackStatus === 'DELETE_IN_PROGRESS') {
+    return true;
+  }
+
+  // Case 3: we're cleaning up after a successful rollforward.
+  // Case 4: we're rolling back a failed update.
+  // In both cases, either the bucket is also being deleted here, or it's just
+  // the CR that's being deleted.
+  // `GetTemplate` will show us the template we are moving to ('new' in case 3,
+  // 'old' in case 4). We will check if the bucket is in the template returned
+  // by `GetTemplate` to see if we need to clean it.
+  const destinationTemplateResponse = await cfn.getTemplate({ StackName: stackId, TemplateStage: 'Processed' }).promise();
+  let template;
+  try {
+    template = yaml.parse(destinationTemplateResponse.TemplateBody ?? '{}', {
```

Comment:
If you [describe the changeset](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudFormation.html#describeChangeSet-property) here instead of analyzing the template then you don't need the `yaml` dependency (and the script to include it in the handler). Better?

Reply:
Maybe. It artificially limits the template's use to cases where we deploy using change sets. Granted, the CDK always does that, but I hate artificially limiting ourselves in that way.